messaging-activemq/src/main/asciidoc/specs/WFLY-5838_artemis_thread_pools.adoc

```diff
@@ -0,0 +1,440 @@
+
+= Non-Requirements
+
+* It is not expected to handle legacy `messaging` subystem migration. If a legacy `messaging` subsystem with configured attributes that relates to thread pools is migrated to the current `messaging-activemq`, the migrate operation will use the corresponding attributes in the `messaging-activemq` subsystem and will *not* create thread pools.
+
+= Design Details
+
+*This design analysis is based on the 2.3.0 release of Apache ActiveMQ Artemis.*
+
+Artemis uses several thread pools:
+
+* server scheduled thread pool - used on the server side for scheduled tasks (e.g. message expiry)
+* server "regular" thread pool - used on the server side for any thread tasks (other than journal I/O)
+* server I/O thread pool - used on the server side for any journal-related task (both file-based and JDBC journals)
+* client scheduled thread pool - used on the client side for scheduled tasks (e.g. network health checks)
+* client "regular" thread pool - used on the client side for any thread tasks
+
+Note that it is common usage to have Artemis clients running inside WildFly (e.g. deployments using JMS, Message-driven Beans).
+
+By default, Artemis uses different types of thread pools for all these cases. We will list them as well as the configuration provided by WildFly's `messaging-activemq` subsystem to modify the default configuration.
+For the first initial release of WildFly using injected thread pools, we propose to inject thread pools configured by WildFly that are *equivalent* to Artemis defaults. Based on that, we may then provide further improvements or recommendations depending on the use cases.
+
+== Artemis Server Thread Pool
+
+By default Artemis configures its server thread pool at https://github.com/apache/activemq-artemis/blob/master/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ActiveMQServerImpl.java#L2182[ActiveMQServerImpl#initializeExecutorServices]
+
+It uses an https://github.com/apache/activemq-artemis/blob/master/artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ActiveMQThreadPoolExecutor.java[ActiveMQThreadPoolExecutor] which is a specialized `ThreadPoolExecutor` that queues tasks if the maximum size is reached instead of rejecting them.
+
+Default parameters are:
+
+----
+* coreSize = 0
+* maximumPoolSize = 30
+* keepAlive = 60 seconds
+* threadFactory (groupName = "ActiveMQ-server-" + server.toString(), dameon = false)
+----
+
+WARNING: We need feedback from Artemis dev team to understand if the use of their ActiveMQThreadPoolExecutor (and its specialized `LinkedBlockingQueue`) is essential to the behaviour of Artemis or if we can circumvent it in WildFly by using a properly configured `LinkedBlockingQueue`. It might be also possible to update WildFly-managed thread pools to offer similar specialization.
```

Comment:
At later versions.. this has been updated:

the implementation now is as simple as this:

```
/*
 * ActiveMQThreadPoolExecutor: a special ThreadPoolExecutor that combines
 * the benefits of a cached executor and a fixed size executor.
 * Similar to a cached executor, threads exceeding the core size are only created on demand,
 * and will be removed after idling for a specified keep time.
 * But in contrast to a standard cached executor, tasks are queued if the
 * maximum pool size if reached, instead of rejected.
 */
public class ActiveMQThreadPoolExecutor extends ThreadPoolExecutor {

   // Handler executed when a task is submitted and a new thread cannot be created (because maxSize was reached)
   // It queues the task on the executors's queue (using the add() method, see ThreadPoolQueue class below)
   private static final RejectedExecutionHandler QUEUE_EXECUTION_HANDLER = (r, e) -> {
      if (!e.isShutdown()) {
         e.getQueue().add(r);
      }
   };

   // A specialized LinkedBlockingQueue that takes new elements by calling add() but not offer()
   // This is to force the ThreadPoolExecutor to always create new threads and never queue
   private static class ThreadPoolQueue extends LinkedBlockingQueue<Runnable> {

      @Override
      public boolean offer(Runnable runnable) {
         return false;
      }

      @Override
      public boolean add(Runnable runnable) {
         return super.offer( runnable );
      }
   }

   public ActiveMQThreadPoolExecutor(int coreSize, int maxSize, long keep, TimeUnit keepUnits, ThreadFactory factory) {
      super( coreSize, maxSize, keep, keepUnits, new ThreadPoolQueue(), factory, QUEUE_EXECUTION_HANDLER );
   }
}

```


However, all we need is a valid thread pool we use with our OrderedExecutorFactory